lxd/instances_put.go

```diff
@@ -0,0 +1,207 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/lxc/lxd/lxd/db"
+	"github.com/lxc/lxd/lxd/instance"
+	"github.com/lxc/lxd/lxd/operations"
+	"github.com/lxc/lxd/lxd/response"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/api"
+	"github.com/lxc/lxd/lxd/cgroup"
+)
+
+
+/*
+* Update state in bulk
+*/
+func containersPut(d *Daemon, r *http.Request) response.Response {
+	project := projectParam(r)
+
+	c, err := instance.LoadByProject(d.State(), project)
+	if err != nil {
+		return response.BadRequest(err)
+	}
+
+	raw := api.InstancesPut{}
+	raw.Timeout = -1
+	if err := json.NewDecoder(r.Body).Decode(&raw); err != nil {
+		return response.BadRequest(err)
+	}
+
+	action := shared.InstanceAction(raw.Action)
+	
+	var names[] string
+	var instances[] instance.Instance
+	for _, inst := range c {
+		switch action {
+		case shared.Start:
+			if inst.IsRunning() {
+				continue
+			}
+		case shared.Stop:
+			if !inst.IsRunning() {
+				continue
+			}
+		}
+
+		instances = append(instances, inst)
+		names = append(names, inst.Name())
+	}
+
+	// Don't mess with containers while in setup mode
+	<-d.readyChan
+
+	var opType db.OperationType
+	var do func(*operations.Operation) error
+	switch action {
+	case shared.Start:
+		opType = db.OperationContainerStart
+		do = func(op *operations.Operation) error {
+			for _, inst := range instances {
+				inst.SetOperation(op)
+				if err = inst.Start(raw.Stateful); err != nil {
+					return err
```

Comment:
How should we convey error state here? If one instance fails to set state we will exit and not set state for any of the others after it. I wasn't sure how to go about "collecting" errors for all these bulk state changes, or how to immediately convey one without breaking iteration.

Reply:
Its a good question, you can certainly collect a slice of errors and then process them at the end (perhaps concatenating them into a single multi-line error).

@stgraber is there any convention around handling "partial" error responses from the API?